boa/src/syntax/parser/cursor/buffered_lexer/mod.rs

```diff
@@ -0,0 +1,212 @@
+    ///         peek(1, false) == 'B'
+    ///         peek(2, false) == \n
+    ///         peek(2, true) == 'C'
+    ///         peek(3, true) == None (End of stream)
+    ///  Note:
+    ///     peek(3, false) == 'C' iff peek(3, true) hasn't been called previously, this is because
+    ///     with skip_line_terminators == true the '\n' would be discarded. This leads to the following statements
+    ///     evaluating to true (in isolation from each other or any other previous cursor calls):
+    ///         peek(3, false) == peek(3, false) == '\n'
+    ///         peek(3, true) == peek(3, true) == None
+    ///         peek(3, true) == peek(3, false) == None
+    ///         (peek(3, false) == 'C') != (peek(3, true) == None)
+    ///
+    /// This behaviour is demonstrated directly in the cursor::tests::skip_peeked_terminators test.
+    ///
+    pub(super) fn peek(
+        &mut self,
+        skip_n: usize,
+        skip_line_terminators: bool,
+    ) -> Result<Option<Token>, ParseError> {
+        let _timer = BoaProfiler::global().start_event("cursor::peek()", "Parsing");
+        if skip_n > MAX_PEEK_SKIP {
+            unimplemented!("peek(n) where n > {}", MAX_PEEK_SKIP);
+        }
+
+        if skip_line_terminators {
+            // We must go through the peeked buffer to remove any line terminators.
+            // This only needs to be done upto the point at which we are peeking - it is
+            // important that we don't go further than this as we would risk removing line terminators
+            // which are later needed.
+            for i in 0..min(skip_n + 1, self.buf_size) {
+                let index = (self.back_index + i) % PEEK_BUF_SIZE;
+                if let Some(t) = self.peeked[index].clone() {
+                    if t.kind() == &TokenKind::LineTerminator {
+                        self.peeked[index].take(); // Remove the line terminator
+
+                        let mut dst_index = index; // Dst index for the swap.
+
+                        // Move all subsequent values up (asif the line terminator never existed).
+                        for j in (i + 1)..self.buf_size {
+                            let src_index = (self.back_index + j) % PEEK_BUF_SIZE; // Src index for the swap.
+                            self.peeked[dst_index] = self.peeked[src_index].take();
+                            dst_index = src_index;
+                        }
+
+                        self.buf_size -= 1;
+                    }
+                }
+            }
+        }
+
+        while self.buf_size <= skip_n {
+            // Need to keep peeking more values.
+            self.peeked[(self.back_index + self.buf_size) % PEEK_BUF_SIZE] =
+                self.lexer.next(skip_line_terminators)?;
+
+            self.buf_size += 1;
+        }
+
+        // Have now peeked ahead the right number of spaces so can fetch the value directly.
+        Ok(self.peeked[(self.back_index + skip_n) % PEEK_BUF_SIZE].clone())
```

Comment:
I wouldn't `clone()` here. A `peek()` should return a reference to the token if the token is not `Copy` (which will probably be with the future string interner).

Reply:
Fair point, 

> There is a bug related to binary operators like `+`, `||` etc where they don't skip line terminators and as result of that the following code fails:
> 
> ```js
> "Hello, " +
> "world"
> ```
> 
> or
> 
> ```js
> "Hello, "
> +
> "world"
> ```
> 
> With the error message:
> 
> ```
> Parsing Error: unexpected token 'line terminator', primary expression at line 1, col 11
> ```
> 
> but this works fine:
> 
> ```js
> "Hello, " + "world"
> ```
> 
> This is the cause for the stack overflow in `test262` branch

This should be a fairly easy fix (I probably shouldn't say that...)